### DISCUSSION TITLE
코드의 실행결과와 이유를 설명해주세요.

### DISCUSSION THREAD

**Discussion by pipisebastian** [publdaze/js-study#75] [실행컨텍스트]
```js
var x = 1;
const y = 1;

console.log(window.x);
console.log(window.y);
console.log(window.z);
console.log(z);
```

```js
function a() {
    console.log("a");
}

const b = () => {
    console.log("b");
}

window.a();
window.b();
```
- 전역에서 let, const로 선언된 변수의 경우 전역객체의 프로퍼티가 되지 않는 이유가 무엇인가요?

**Comment by pipisebastian** [comment]
- 1번째 코드 : 1, undefined, undefined, ReferenceError
  - window.x의 경우, var키워드로 선언한 x가 전역객체(window)의 프로퍼티로 들어가기에 1이 출력이 됩니다.
  - y는 전역코드에 선언되긴 하지만, let 키워드로 선언하여, **전역객체와 연결되지  않는 선언적 환경 레코드에 들어가게 됩니다**. 그렇기에 undefined가 반환됩니다.
  -  객체의 속성에 접근할 때 해당 속성이 존재하지 않으면 undefined가 반환
  - 존재하지 않는 식별자 → ReferenceError가 발생

- 2번째 코드 : a, TypeError(window.b is not a function)
  - 함수 선언식인 a는 객체환경레코드에 들어가, 전역객체의 메서드가 되기에 정상적으로 동작하나, b의 경우, 선언적 환경 레코드에 들어갑니다. 그렇기에 전역객체에서 참조할 수 없게 됩니다.

**Comment by publdaze** [comment]
> - 코드의 실행결과와 이유를 설명해주세요.
> 1.
>   ```js
>   var x = 1;
>   const y = 1;
>   
>   console.log(window.x);
>   console.log(window.y);
>   console.log(window.z);
>   console.log(z);
>   ```
> 2.
>   ```js
>   function a() {
>       console.log("a");
>   }
>   
>   const b = () => {
>       console.log("b");
>   }
>   
>   window.a();
>   window.b();
>   ```
### 1번 코드의 경우
- x는 var로 선언이 되어 있고, y는 const로 선언되어 있음
- 전역 렉시컬 환경에서 var로 선언된 변수 식별자는 객체 환경 레코드에 바인딩된 전역 객체의 프로퍼티로 들어가고,
- const로 선언된 변수 식별자는 선언적 환경 레코드에서 관리 됨
- 브라우저의 경우 전역 객체는 window로 호출이 가능하기 때문에
- 첫번째 출력에서는 x에 할당된 1이 출력이 되고,
- 두번째 출력에서 window에서는 y라는 프로퍼티가 존재하지 않기 때문에 undefined가 출력됨
- 세번째 출력에서는 두번째 출력과 마찬가지로 window 객체에 z라는 프로퍼티가 존재하지 않아 undefined가 출력되고
- 네번째 출력에는 전역 렉시컬 환경에서 z라는 식별자를 검색할 수 없기 때문에 reference에러가 발생함
### 2번 코드의 경우
- a는 함수선언문이고, b는 함수 표현식으로 초기화 됨
- a는 1번 코드의 var와 마찬가지로 전역객체의 메서드로 들어가게 되고
- b의 경우 평가단계에서는 const로 선언된 변수 선언으로 선언적 환경 레코드에서 관리 됨
- 따라서 첫번째 출력에는 a가 정상적으로 호출되어 "a"를 출력하지만,
- 두번째 출력에는 window 객체에 b라는 메소드가 존재하지 않아 b를 호출할 수 없다는 type에러가 발생하게 됨
> - 전역에서 let, const로 선언된 변수의 경우 전역객체의 프로퍼티가 되지 않는 이유가 무엇인가요?
- let, const는 독자적인 블록레벨 스코프를 가지며, 선언적 환경 레코드에서 관리되기 때문
